typescript/trans/typechecking/typescript.nabl2

```diff
@@ -253,7 +255,36 @@ rules
       Return{"function"} <- function_scope,
       Return{"function"} : returnTy,
       Map1 [[ body ^ (function_scope) ]].
-     
+    
+    
+    [[ ClassDeclaration(None(), _, parent, members) ^ (s) ]] :=
+      new clazz,
+      clazz -P-> s,
+      Map1 [[ members ^ (clazz) ]],
+      ClassParent [[ parent ^ (s): parentTy ]].
+      
+    [[ ClassDeclaration(Some(name), _, parent, members) ^ (s) ]] :=
+      new clazz,
+      clazz -P-> s,
+      Value{name} <- s,
+      classTy == RECORD(clazz),
+      Value{name} : classTy !,
+      Type{name} <- s,
+      Type{name} : classTy !,
+      Value{"this"} <- clazz,
+      Value{"this"} : classTy !,
+      ClassParent [[ parent ^ (s): parentTy ]],
+      classTy <! parentTy,
+      Map1 [[ members ^ (clazz) ]].
+    
+    ClassParent [[ None() ^ (s): RECORD(s) ]] := true.
+    ClassParent [[ Some(ClassExtends(parent)) ^ (s): parentTy ]] :=
+       [[ parent ^ (s): parentTy ]],
+       parentTy == RECORD(parentScope) | error $[Expected class as parent, but got [ty]],
+       Value{"this"} -> parentScope,
```

Comment:
So... we would need to be able to distinguish classes from records because we cannot assign records to a class, but this is basically impossible because the class parent can be in `interface`, which is a record. amirite?